ouroboros-consensus/test-consensus/Test/Dynamic/LeaderSchedule.hs

```diff
@@ -62,47 +64,79 @@ prop_simple_leader_schedule_convergence :: NumSlots
                                         -> Seed
                                         -> Property
 prop_simple_leader_schedule_convergence numSlots numCoreNodes params seed =
+  forAllShrink
+    (genConnectionPlan numCoreNodes numSlots)
+    shrinkConnectionPlan_keepNodes
+  $ \connPlan ->
     forAllShrink
-        (genLeaderSchedule numSlots numCoreNodes params)
-        (shrinkLeaderSchedule numSlots)
+        (genLeaderSchedule k connPlan)
+        (shrinkLeaderSchedule k connPlan)
         $ \schedule ->
-            let longest = longestCrowdedRun schedule
+            let longest = longestCrowdedRun connPlan schedule
             in    counterexample ("schedule: " <> condense schedule <> "\n" <> show longest)
-                $ label ("longest crowded run " <> show (crowdedRunLength longest))
-                $ prop_simple_protocol_convergence
+                $ label ("longest crowded run " <> show (maybe 0 crowdedRunLength longest))
+                $ prop_simple_protocol_convergence'
                     (\nid -> protocolInfo numCoreNodes nid (ProtocolLeaderSchedule params schedule))
                     isValid
-                    numCoreNodes
-                    numSlots
+                    connPlan
                     seed
   where
+    k = praosSecurityParam params
+
     isValid :: [NodeId]
-            -> Map NodeId ( NodeConfig ProtocolLeaderSchedule
-                          , Chain (SimplePraosRuleBlock SimpleMockCrypto)
-                          )
+            -> ChainsTrace SimpleMockCrypto SimplePraosRuleExt
             -> Property
-    isValid nodeIds final =
+    isValid nodeIds ct =
             counterexample (tracesToDot final)
+       $    counterexample prettySwitches
        $    tabulate "shortestLength"
-            [show (rangeK (praosSecurityParam params) (shortestLength final'))]
+            [show (rangeK k (shortestLength final'))]
        $    Map.keys final === nodeIds
        .&&. prop_all_common_prefix
-              (maxRollbacks $ praosSecurityParam params)
+              (maxRollbacks k)
               (Map.elems final')
       where
+        final = ctFinalChains ct
+        switches = ctSwitches ct
+
         -- Without the 'NodeConfig's
         final' = snd <$> final
 
+        prettySwitches =
+            unlines $
+            map condense $
+            [   (nid, slot, condenseChainEvent ev)
+            | (nid, slot, ev) <- reverse switches
+            ]
+
+condenseChainEvent :: TraceChainEvent SimpleMockCrypto SimplePraosRuleExt -> String
+condenseChainEvent = \case
+  SwitchedTo af -> "SwitchedTo " ++ condenseAF af
```

Comment:
When you have a `where`, indent the lines before it too, e.g.,

```haskell
condenseChainEvent = \case
    SwitchedTo af -> "SwitchedTo " <> condenseAF af
    Forged blk    -> "Forged "     <> condense blk
  where
    condenseAF ..

Reply:
Done